File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleLaunchArgs.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.Immutable
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleFlowRoute
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleProduct
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleScenario
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleThemeScenario

/** How long the host holds the camera before handing off, so the SDK meets a contended device. */
sealed interface UseSmileIDSampleHoldCamera {
    data object Keep : UseSmileIDSampleHoldCamera

    data class Millis(val value: Long) : UseSmileIDSampleHoldCamera
}

/** Rendered in the argument's own vocabulary, so a report cannot drift from what the parser accepts. */
fun UseSmileIDSampleHoldCamera.describe(): String = when (this) {
    UseSmileIDSampleHoldCamera.Keep -> "holdCamera=${UseSmileIDSampleLaunchArgs.HOLD_CAMERA_KEEP}"
    is UseSmileIDSampleHoldCamera.Millis -> "holdCamera=${value}ms"
}

/** The canonical arguments from `spec/launch-args.json`; reading an Intent is the shell's job. */
@Immutable
data class UseSmileIDSampleLaunchArgs(
    val scenario: UseSmileIDSampleScenario = UseSmileIDSampleScenario.Normal,
    val theme: UseSmileIDSampleThemeScenario = UseSmileIDSampleThemeScenario.BrandDefault,
    val route: UseSmileIDSampleFlowRoute = UseSmileIDSampleFlowRoute.Fullscreen,
    val autostart: UseSmileIDSampleProduct? = null,
    /** Automation precondition only — see `spec/launch-args.json`. */
    val seedJobs: Boolean = false,
    /** The design's three profiles instead of the one empty starter; in memory, so per launch. */
    val seedProfiles: Boolean = false,
    /** Reveals the result card on a release build. Always on in debug, so only a release run needs it. */
    val probes: Boolean = false,
    val appLocale: String? = null,
    val holdCamera: UseSmileIDSampleHoldCamera? = null,
    /** Seconds a transient notice stays — see `spec/launch-args.json`. Automation only. */
    val noticeWindow: Int? = null,
) {
    companion object {
        const val SCENARIO = "scenario"
        const val THEME = "theme"
        const val ROUTE = "route"
        const val AUTOSTART = "autostart"
        const val SEED_JOBS = "seedJobs"
        const val SEED_PROFILES = "seedProfiles"
        const val PROBES = "probes"
        const val APP_LOCALE = "appLocale"
        const val HOLD_CAMERA = "holdCamera"
        const val NOTICE_WINDOW = "noticeWindow"

        val names =
            listOf(SCENARIO, THEME, ROUTE, AUTOSTART, SEED_JOBS, SEED_PROFILES, PROBES, APP_LOCALE, HOLD_CAMERA, NOTICE_WINDOW)

        internal const val HOLD_CAMERA_KEEP = "keep"

        /** An unrecognised value falls back to its default, which is safe only because the card reports it. */
        fun from(raw: Map<String, Any?>): UseSmileIDSampleLaunchArgs {
            val defaults = UseSmileIDSampleLaunchArgs()
            return UseSmileIDSampleLaunchArgs(
                scenario = UseSmileIDSampleScenario.entries.firstOrNull { it.id == raw.string(SCENARIO) }
                    ?: defaults.scenario,
                theme = UseSmileIDSampleThemeScenario.entries.firstOrNull { it.id == raw.string(THEME) }
                    ?: defaults.theme,
                route = UseSmileIDSampleFlowRoute.entries.firstOrNull { it.id == raw.string(ROUTE) }
                    ?: defaults.route,
                autostart = UseSmileIDSampleProduct.entries.firstOrNull { it.id == raw.string(AUTOSTART) },
                seedJobs = raw.boolean(SEED_JOBS) ?: defaults.seedJobs,
                seedProfiles = raw.boolean(SEED_PROFILES) ?: defaults.seedProfiles,
                probes = raw.boolean(PROBES) ?: defaults.probes,
                appLocale = raw.string(APP_LOCALE),
                holdCamera = raw.holdCamera(),
                noticeWindow = raw.noticeWindow(),
            )
        }

        /** Positive seconds; anything else falls back to the product's own window. */
        private fun Map<String, Any?>.noticeWindow(): Int? =
            string(NOTICE_WINDOW)?.toIntOrNull()?.takeIf { it > 0 }

        private fun Map<String, Any?>.string(name: String): String? =
            this[name]?.toString()?.trim()?.takeIf { it.isNotEmpty() }

        /** `am start` sends `--ez` as a Boolean and `-e` as a String, and automation uses both. */
        private fun Map<String, Any?>.boolean(name: String): Boolean? = when (val value = this[name]) {
            is Boolean -> value
            else -> value?.toString()?.trim()?.lowercase()?.toBooleanStrictOrNull()
        }

        private fun Map<String, Any?>.holdCamera(): UseSmileIDSampleHoldCamera? {
            val value = string(HOLD_CAMERA) ?: return null
            if (value.equals(HOLD_CAMERA_KEEP, ignoreCase = true)) return UseSmileIDSampleHoldCamera.Keep
            return value.toLongOrNull()
                ?.takeIf { it > 0 }
                ?.let { UseSmileIDSampleHoldCamera.Millis(it) }
        }
    }
}

File: flutter/sample_ui/lib/src/state/use_smileid_sample_launch_args.dart
import 'package:flutter/foundation.dart';

import '../model/use_smileid_sample_product.dart';
import '../model/use_smileid_sample_result.dart';
import '../model/use_smileid_sample_scenario.dart';

/// How long the host holds the camera before handing off, so the SDK meets a contended device.
@immutable
sealed class UseSmileIDSampleHoldCamera {
  const UseSmileIDSampleHoldCamera();
}

/// Hold the camera for the whole run.
@immutable
final class UseSmileIDSampleHoldCameraKeep extends UseSmileIDSampleHoldCamera {
  /// Takes nothing: 'keep' is the whole instruction.
  const UseSmileIDSampleHoldCameraKeep();

  @override
  bool operator ==(Object other) => other is UseSmileIDSampleHoldCameraKeep;

  @override
  int get hashCode => (UseSmileIDSampleHoldCameraKeep).hashCode;
}

/// Hold the camera for a bounded number of milliseconds.
@immutable
final class UseSmileIDSampleHoldCameraMillis extends UseSmileIDSampleHoldCamera {
  /// Takes positive milliseconds; the parser rejects anything else.
  const UseSmileIDSampleHoldCameraMillis(this.value);

  /// The hold, in milliseconds.
  final int value;

  @override
  bool operator ==(Object other) =>
      other is UseSmileIDSampleHoldCameraMillis && other.value == value;

  @override
  int get hashCode => value.hashCode;
}

/// The canonical arguments from `spec/launch-args.json`; reading the cold-start link is the shell's job.
@immutable
class UseSmileIDSampleLaunchArgs {
  /// Every default is the plain launch: no fixtures, no probes, no autostart.
  const UseSmileIDSampleLaunchArgs({
    this.scenario = UseSmileIDSampleScenario.normal,
    this.theme = UseSmileIDSampleThemeScenario.brandDefault,
    this.route = UseSmileIDSampleFlowRoute.fullscreen,
    this.autostart,
    this.seedJobs = false,
    this.seedProfiles = false,
    this.probes = false,
    this.appLocale,
    this.holdCamera,
    this.noticeWindow,
  });

  /// The flow scenario.
  final UseSmileIDSampleScenario scenario;

  /// The theme scenario.
  final UseSmileIDSampleThemeScenario theme;

  /// Which entry route hosts the SDK flow.
  final UseSmileIDSampleFlowRoute route;

  /// A product to launch straight into, skipping the pre-flow forms.
  final UseSmileIDSampleProduct? autostart;

  /// Automation precondition only — seeds the design's fixture verifications.
  final bool seedJobs;

  /// The design's three profiles instead of the one empty starter; in memory, so per launch.
  final bool seedProfiles;

  /// Reveals the result card on a release build; always on in debug.
  final bool probes;

  /// An in-app locale override, because a device locale is not reliably scriptable.
  final String? appLocale;

  /// Whether the host acquires the camera before handing off.
  final UseSmileIDSampleHoldCamera? holdCamera;

  /// Seconds a transient notice stays before dismissing itself.
  final int? noticeWindow;

  /// The `scenario` argument's canonical name.
  static const String scenarioArg = 'scenario';

  /// The `theme` argument's canonical name.
  static const String themeArg = 'theme';

  /// The `route` argument's canonical name.
  static const String routeArg = 'route';

  /// The `autostart` argument's canonical name.
  static const String autostartArg = 'autostart';

  /// The `seedJobs` argument's canonical name.
  static const String seedJobsArg = 'seedJobs';

  /// The `seedProfiles` argument's canonical name.
  static const String seedProfilesArg = 'seedProfiles';

  /// The `probes` argument's canonical name.
  static const String probesArg = 'probes';

  /// The `appLocale` argument's canonical name.
  static const String appLocaleArg = 'appLocale';

  /// The `holdCamera` argument's canonical name.
  static const String holdCameraArg = 'holdCamera';

  /// The `noticeWindow` argument's canonical name.
  static const String noticeWindowArg = 'noticeWindow';

  /// The value `holdCamera` takes to hold for the whole run.
  static const String holdCameraKeep = 'keep';

  /// The ten names in the spec's own order, which the spec test compares against.
  static const List<String> names = <String>[
    scenarioArg,
    themeArg,
    routeArg,
    autostartArg,
    seedJobsArg,
    seedProfilesArg,
    probesArg,
    appLocaleArg,
    holdCameraArg,
    noticeWindowArg,
  ];

  /// An unrecognised value falls back to its default, which is safe only because the card reports it.
  static UseSmileIDSampleLaunchArgs from(Map<String, Object?> raw) {
    const UseSmileIDSampleLaunchArgs defaults = UseSmileIDSampleLaunchArgs();
    return UseSmileIDSampleLaunchArgs(
      scenario: UseSmileIDSampleScenario.byId(_string(raw, scenarioArg)) ?? defaults.scenario,
      theme: UseSmileIDSampleThemeScenario.byId(_string(raw, themeArg)) ?? defaults.theme,
      route: UseSmileIDSampleFlowRoute.byId(_string(raw, routeArg)) ?? defaults.route,
      autostart: UseSmileIDSampleProduct.byId(_string(raw, autostartArg)),
      seedJobs: _boolean(raw, seedJobsArg) ?? defaults.seedJobs,
      seedProfiles: _boolean(raw, seedProfilesArg) ?? defaults.seedProfiles,
      probes: _boolean(raw, probesArg) ?? defaults.probes,
      appLocale: _string(raw, appLocaleArg),
      holdCamera: _holdCamera(raw),
      noticeWindow: _noticeWindow(raw),
    );
  }

  /// Everything a cold-start link carries, read from its query string.
  static UseSmileIDSampleLaunchArgs fromUri(Uri? uri) =>
      from(<String, Object?>{...?uri?.queryParameters});

  static String? _string(Map<String, Object?> raw, String name) {
    final String value = raw[name]?.toString().trim() ?? '';
    return value.isEmpty ? null : value;
  }

  /// A link's query parameter arrives as a String, which is the only form that path has.
  static bool? _boolean(Map<String, Object?> raw, String name) {
    final Object? value = raw[name];
    if (value is bool) {
      return value;
    }
    final String? text = _string(raw, name)?.toLowerCase();
    return switch (text) {
      'true' => true,
      'false' => false,
      _ => null,
    };
  }

  /// Null rather than zero for a rejected value: zero would dismiss the notice before it rendered.
  static int? _noticeWindow(Map<String, Object?> raw) {
    final int? seconds = int.tryParse(_string(raw, noticeWindowArg) ?? '');
    return seconds != null && seconds > 0 ? seconds : null;
  }

  static UseSmileIDSampleHoldCamera? _holdCamera(Map<String, Object?> raw) {
    final String? value = _string(raw, holdCameraArg);
    if (value == null) {
      return null;
    }
    if (value.toLowerCase() == holdCameraKeep) {
      return const UseSmileIDSampleHoldCameraKeep();
    }
    final int? millis = int.tryParse(value);
    return millis != null && millis > 0 ? UseSmileIDSampleHoldCameraMillis(millis) : null;
  }

  @override
  bool operator ==(Object other) =>
      other is UseSmileIDSampleLaunchArgs &&
      other.scenario == scenario &&
      other.theme == theme &&
      other.route == route &&
      other.autostart == autostart &&
      other.seedJobs == seedJobs &&
      other.seedProfiles == seedProfiles &&
      other.probes == probes &&
      other.appLocale == appLocale &&
      other.holdCamera == holdCamera &&
      other.noticeWindow == noticeWindow;

  @override
  int get hashCode => Object.hash(
    scenario,
    theme,
    route,
    autostart,
    seedJobs,
    seedProfiles,
    probes,
    appLocale,
    holdCamera,
    noticeWindow,
  );
}
